### PR TITLE
[flutter_tools] unpin SDK deps when upgrading packages

### LINKED ISSUE #36511: flutter update-packages --force-upgrade fails to un-pin flutter sdk deps
## Steps to reproduce

1. Create the following `pubspec.yaml` file in a clean directory (henceforth referred to a `$DIR`):

    ```yaml
    name: bug_report
    dependencies:
      flutter:
        sdk: flutter
      meta: any
    ```

2. Update your `$FLUTTER_ROOT/packages/flutter/pubspec.yaml` file to specify a pinned version of `meta: 1.1.6` (which is what it's [currently pinned to](https://github.com/flutter/flutter/blob/82a4ba40bb59cebde169c58253599a91495b376d/packages/flutter/pubspec.yaml#L13))

3. Run the following commands:

    ```sh
    $ cd $DIR
    $ FLUTTER_ROOT=/path/to/flutter_root \
        PUB_ENVIRONMENT=flutter_cli \
        pub upgrade
    ```

## Observed results

Because your pubspec.yaml depends on the Flutter SDK, it's pinned to the same version of `meta` that Flutter is pinned to.  This is working as intended.

However, the previous steps are _exactly_ what `flutter update-packages --force-upgrade` does, with the exception of putting the Flutter SDK dependency in its synthetic `pubspec.yaml` file.  Unfortunately, some of Flutter's transitive deps require the Flutter SDK and thus are triggering this behavior.  The net result is that any non-dev dependency listed in the Flutter framework's pubspec is pinned and never upgraded.

### LINKED ISSUE #53350: Not all pinned dependencies upgrade with `flutter update-packages`
Flutter currently pins several packages to outdated versions:
```
$ cd packages/flutter
$ flutter pub outdated
Package           Current   Upgradable  Resolvable  Latest
dependencies
collection        *1.14.11  *1.14.11    *1.14.11    1.14.12

dev_dependencies
archive           *2.0.11   *2.0.11     *2.0.11     2.0.13
args              *1.5.2    *1.5.2      *1.5.2      1.6.0
async             *2.4.0    *2.4.0      *2.4.0      2.4.1
boolean_selector  *1.0.5    *1.0.5      *1.0.5      2.0.0
charcode          *1.1.2    *1.1.2      *1.1.2      1.1.3
crypto            *2.1.3    *2.1.3      *2.1.3      2.1.4
image             *2.1.4    *2.1.4      *2.1.4      2.1.12
petitparser       *2.4.0    *2.4.0      *2.4.0      3.0.1
quiver            *2.0.5    *2.0.5      *2.0.5      2.1.3
source_span       *1.5.5    *1.5.5      *1.5.5      1.7.0
xml               *3.5.0    *3.5.0      *3.5.0      3.7.0

transitive dependencies: all up-to-date
```

I tried to updated these with `flutter update-packages --force-upgrade` in https://github.com/flutter/flutter/pull/53349, however that only moved one of them (`http_parser: 3.1.3` -> `3.1.4`). Is there a bug in `flutter update-packages --force-upgrade` ?

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

When run update-packages --force-upgrade, tell pub to use a flutter SDK with only `any` constraints. This prevents the prior version of the flutter SDK pubspecs from affecting the update packages version solve.

Fixes https://github.com/flutter/flutter/issues/53350
Fixes https://github.com/flutter/flutter/issues/36511

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/update_packages.dart [hunk 1]
```diff
@@ -1334,3 +1351,77 @@ String _computeChecksum(Iterable<String> names, String getVersion(String name))
   }
   return ((upperCheck << 8) | lowerCheck).toRadixString(16).padLeft(4, '0');
 }
+
+/// Create a synthetic Flutter SDK so that pub version solving does not get
+/// stuck on the old versions.
+Directory createTemporaryFlutterSdk(FileSystem fileSystem, Directory realFlutter, List<PubspecYaml> pubspecs) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/update_packages.dart]: So this is filling in a fake Flutter SDK to trick pub into doing the right thing. How will we know if something changes with pub, and this no longer can trick it?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/update_packages.dart]: I wouldn't say its tricking pub, its exposing the SDK the same way we do for `flutter packages get`, via FLUTTER_ROOT and the file system. If the implementation of pub changes enough to break this, everything else regarding dependency solving with the SDK will break too

File: packages/flutter_tools/lib/src/commands/update_packages.dart [hunk 2]
```diff
@@ -276,14 +276,30 @@ class UpdatePackagesCommand extends FlutterCommand {
         final File fakePackage = _pubspecFor(tempDir);
         fakePackage.createSync();
         fakePackage.writeAsStringSync(_generateFakePubspec(dependencies.values));
-        // First we run "pub upgrade" on this generated package:
+        // First, we create a synthetic flutter SDK so that transitive flutter SDK
+        // constraints are not effected by this upgrade.
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/update_packages.dart]: Fixed

File: packages/flutter_tools/lib/src/commands/update_packages.dart (hunk 1 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/update_packages.dart]: Fair enough

**Review by zanderso — APPROVED** [review]

**Comment by mit-mit** [comment]
Thanks much @jonahwilliams !

**Pull request merged**